whistl/train.py

```diff
@@ -0,0 +1,480 @@
+                # of the others, and the theoretical basis can be found in the Invariant
+                # Risk Minimization paper
+                penalty = compute_irm_penalty(loss, dummy_w)
+                train_penalty += penalty.item()
+
+                optimizer.zero_grad()
+                # Calculate the gradient of the combined loss function
+                train_loss += float(loss_scaling_factor * loss + penalty)
+                (loss_scaling_factor * loss + penalty).backward(retain_graph=False)
+                optimizer.step()
+
+            tune_loss = 0
+            tune_correct = 0
+            # Speed up validation by telling torch not to worry about computing gradients
+            with torch.no_grad():
+                for tune_batch in tune_loader:
+                    expression, labels, ids = tune_batch
+                    tune_expression = expression.float().to(device)
+                    tune_labels = labels.to(device).float()
+
+                    loss_function = nn.BCEWithLogitsLoss()
+
+                    tune_preds = classifier(tune_expression)
+                    loss = loss_function(tune_preds, tune_labels)
+                    tune_loss += loss.item()
+                    tune_correct += utils.count_correct(tune_preds, tune_labels)
+
+                # Save the model
+                if save_file is not None:
+                    if best_tune_loss is None or tune_loss < best_tune_loss:
+                        best_tune_loss = tune_loss
+                        if epoch > burn_in_epochs:
+                            torch.save(classifier, save_file)
+
+            tune_loss = tune_loss / len(tune_dataset)
+            tune_acc = tune_correct / len(tune_dataset)
+            train_loss = train_loss / train_sample_count
+            train_acc = train_correct / train_sample_count
+            # We cast these to floats to avoid having to pickle entire Tensor objects
+            train_penalty = float(train_penalty / train_sample_count)
+            train_raw_loss = float(train_raw_loss / train_sample_count)
+
+            results['train_loss'].append(train_loss)
+            results['train_acc'].append(train_acc)
+            results['tune_loss'].append(tune_loss)
+            results['tune_acc'].append(tune_acc)
+            results['train_penalty'].append(train_penalty)
+            results['train_raw_loss'].append(train_raw_loss)
+
+            if logger is not None:
+                logger.info('Epoch {}'.format(epoch))
+                logger.info('Train loss: {}'.format(train_loss))
+                logger.info('Tune loss: {}'.format(tune_loss))
+                logger.info('Train accuracy: {}'.format(train_acc))
+                logger.info('Tune accuracy: {}'.format(tune_acc))
+                logger.info('Baseline accuracy: {}'.format(baseline))
+
+    except Exception as e:
+        logger.error(e, exc_info=True)
+    finally:
+        # results = utils.add_genes_to_results(results, gene_file)
```

Comment:
Do you want this commented out, or can it be removed?

Reply:
I'll go ahead and remove it. I don't think I'll need to add it back in the future